lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1435,6 +1456,1086 @@ class BuildIfGroups : public calyx::FuncOpPartialLoweringPattern {
+      auto *banks = jsonObj->getArray("banks");
+      uint availableBanks = 0;
+      if (auto bankOpt = (*banks)[memNum.at(origMemDef)].getAsInteger())
+        availableBanks = *bankOpt;
+      else {
+        std::string dumpStr;
+        llvm::raw_string_ostream dumpStream(dumpStr);
+        origMemDef->print(dumpStream);
+        report_fatal_error(
+            llvm::Twine(
+                "Cannot find the number of banks associated with memory") +
+            dumpStream.str());
+      }
+      // Allocate new banks for `memOp` into `block`
+      auto allocateBanks = [&](Operation *memOp, MemRefType memTy,
+                               Block *insertBlock) -> SmallVector<Operation *> {
+        auto origShape = memTy.getShape();
+        if (origShape.front() % availableBanks != 0)
+          memOp->emitError()
+              << "memory shape must be divisible by the banking factor";
+        assert(origShape.size() == 1 &&
+               "memref must be flattened before scf-to-calyx pass");
+        uint bankSize = origShape.front() / availableBanks;
+        SmallVector<Operation *> banks;
+        OpBuilder builder = OpBuilder::atBlockBegin(insertBlock);
+        TypeSwitch<Operation *>(origMemDef)
+            .Case<memref::AllocOp>([&](memref::AllocOp allocOp) {
+              for (uint bankCnt = 0; bankCnt < availableBanks; bankCnt++) {
+                auto bankAllocOp = builder.create<memref::AllocOp>(
+                    insertBlock->getParentOp()->getLoc(),
+                    MemRefType::get(bankSize, memTy.getElementType(),
+                                    memTy.getLayout(), memTy.getMemorySpace()));
+                banks.push_back(bankAllocOp);
+              }
+            })
+            .Case<memref::GetGlobalOp>([&](memref::GetGlobalOp getGlobalOp) {
+              OpBuilder::InsertPoint globalOpsInsertPt, getGlobalOpsInsertPt;
+              for (uint bankCnt = 0; bankCnt < availableBanks; bankCnt++) {
+                auto *symbolTableOp =
+                    getGlobalOp->getParentWithTrait<OpTrait::SymbolTable>();
+                auto globalOp = dyn_cast_or_null<memref::GlobalOp>(
+                    SymbolTable::lookupSymbolIn(symbolTableOp,
+                                                getGlobalOp.getNameAttr()));
+                MemRefType type = globalOp.getType();
+                assert(
+                    circt::isUniDimensional(type) &&
+                    "GlobalOp must be flattened before the scf-to-calyx pass");
+                auto cstAttr = llvm::dyn_cast_or_null<DenseElementsAttr>(
+                    globalOp.getConstantInitValue());
+                uint beginIdx = bankSize * bankCnt;
+                uint endIdx = bankSize * (bankCnt + 1);
+                auto allAttrs = cstAttr.getValues<Attribute>();
+                SmallVector<Attribute, 8> extractedElements;
+                for (auto idx = beginIdx; idx < endIdx; idx++)
```

Comment:
Should be:
```suggestion
                for (uint i = 0; i < bankSize; i++) {
                  uint idx = bankCnt + availableBanks * i;
                  extractedElements.push_back(allAttrs[idx]);
                }
```
for round robin